SSH will not detect connection termination

I continue to get hard-to-reproduce lock-ups while executing remote commands via SSH (and also with WinRM #118).

Basically it looks like that the tcp connection get dropped by the remote endpoint (maybe a fascist firewall between the hosts?), but overthere will not detect it and wait forever.

This is the stacktrace got while interrupting a stuck thread:

```
java.io.InterruptedIOException: null
        at net.schmizz.sshj.connection.channel.ChannelInputStream.read(ChannelInputStream.java:130) ~[sshj-0.8.1.jar:na]
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:283) ~[na:1.7.0_65]
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:325) ~[na:1.7.0_65]
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:177) ~[na:1.7.0_65]
        at sun.nio.cs.StreamDecoder.read0(StreamDecoder.java:126) ~[na:1.7.0_65]
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:112) ~[na:1.7.0_65]
        at java.io.InputStreamReader.read(InputStreamReader.java:168) ~[na:1.7.0_65]
        at com.xebialabs.overthere.spi.BaseOverthereConnection$1.run(BaseOverthereConnection.java:328) ~[overthere-2.4.4.jar:na]
Caused by: java.lang.InterruptedException: null
        at java.lang.Object.wait(Native Method) ~[na:1.7.0_65]
        at java.lang.Object.wait(Object.java:503) ~[na:1.7.0_65]
        at net.schmizz.sshj.connection.channel.ChannelInputStream.read(ChannelInputStream.java:128) ~[sshj-0.8.1.jar:na]
        ... 7 common frames omitted
08:49:15.687 [stderr reader] ERROR c.x.o.spi.BaseOverthereConnection - An exception occured reading stderr while executing [chef-solo --no-color] on Thread[stderr reader,5,main]
java.io.InterruptedIOException: null
        at net.schmizz.sshj.connection.channel.ChannelInputStream.read(ChannelInputStream.java:130) ~[sshj-0.8.1.jar:na]
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:283) ~[na:1.7.0_65]
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:325) ~[na:1.7.0_65]
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:177) ~[na:1.7.0_65]
        at sun.nio.cs.StreamDecoder.read0(StreamDecoder.java:126) ~[na:1.7.0_65]
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:112) ~[na:1.7.0_65]
        at java.io.InputStreamReader.read(InputStreamReader.java:168) ~[na:1.7.0_65]
        at com.xebialabs.overthere.spi.BaseOverthereConnection$1.run(BaseOverthereConnection.java:317) ~[overthere-2.4.4.jar:na]
Caused by: java.lang.InterruptedException: null
        at java.lang.Object.wait(Native Method) ~[na:1.7.0_65]
        at java.lang.Object.wait(Object.java:503) ~[na:1.7.0_65]
        at net.schmizz.sshj.connection.channel.ChannelInputStream.read(ChannelInputStream.java:128) ~[sshj-0.8.1.jar:na]
        ... 7 common frames omitted
```

Any hint?

--- a/README.md
+++ b/README.md
@@ -367,6 +367,10 @@ The SSH protocol implementation of Overthere defines a number of additional conn
 	</td>
 </tr>
 <tr>
+	<th align="left" valign="top"><a name="heartbeatInterval"></a>heartbeatInterval</th>
+	<td>Specify an interval to send keep-alives packets. Default is 0 (no keep-alive).</td>
+</tr>
+<tr>
 	<th align="left" valign="top"><a name="ssh_interactiveKeyboardAuthRegex"></a>interactiveKeyboardAuthRegex</th>
 	<td>The regular expression to look for in keyboard-interactive prompts before sending the password. The default value is <code>.*Password:[ ]?</code>. When the SSH server is configured to not allow <a href="http://www.ietf.org/rfc/rfc4252.txt">password authentication</a> but is configured to allow <a href="http://www.ietf.org/rfc/rfc4256.txt">keyboard-interactive authentication</a> using passwords, Overthere will compare the interactive-keyboard prompt against this regular expression and send the value of the <strong>password</strong> option when they match. The default value is <code>.*Password:[ ]?</code></td>
 </tr>

--- a/src/main/java/com/xebialabs/overthere/ssh/SshConnection.java
+++ b/src/main/java/com/xebialabs/overthere/ssh/SshConnection.java
@@ -57,6 +57,7 @@ import com.xebialabs.overthere.RuntimeIOException;
 import com.xebialabs.overthere.spi.AddressPortMapper;
 import com.xebialabs.overthere.spi.BaseOverthereConnection;
 
+import static com.xebialabs.overthere.ssh.SshConnectionBuilder.*;
 import static com.xebialabs.overthere.util.OverthereUtils.checkArgument;
 import static com.xebialabs.overthere.util.OverthereUtils.checkNotNull;
 import static com.xebialabs.overthere.util.OverthereUtils.checkState;
@@ -64,19 +65,6 @@ import static com.xebialabs.overthere.ConnectionOptions.ADDRESS;
 import static com.xebialabs.overthere.ConnectionOptions.PASSWORD;
 import static com.xebialabs.overthere.ConnectionOptions.PORT;
 import static com.xebialabs.overthere.ConnectionOptions.USERNAME;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.ALLOCATE_DEFAULT_PTY;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.ALLOCATE_DEFAULT_PTY_DEFAULT;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.ALLOCATE_PTY;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.CONNECTION_TYPE;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.INTERACTIVE_KEYBOARD_AUTH_PROMPT_REGEX;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.INTERACTIVE_KEYBOARD_AUTH_PROMPT_REGEX_DEFAULT;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.LOCAL_ADDRESS;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.LOCAL_PORT;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.OPEN_SHELL_BEFORE_EXECUTE;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.OPEN_SHELL_BEFORE_EXECUTE_DEFAULT;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.PASSPHRASE;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.PORT_DEFAULT_SSH;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.PRIVATE_KEY_FILE;
 import static com.xebialabs.overthere.util.OverthereUtils.closeQuietly;
 import static com.xebialabs.overthere.util.OverthereUtils.constructPath;
 import static java.lang.String.format;
@@ -117,6 +105,8 @@ abstract class SshConnection extends BaseOverthereConnection {
 
     protected String allocatePty;
 
+    protected int heartbeatInterval;
+
     protected SSHClient sshClient;
 
     private static final Pattern ptyPattern = Pattern.compile(PTY_PATTERN);
@@ -146,6 +136,7 @@ abstract class SshConnection extends BaseOverthereConnection {
         privateKeyFile = options.getOptional(PRIVATE_KEY_FILE);
         passphrase = options.getOptional(PASSPHRASE);
         allocateDefaultPty = options.getBoolean(ALLOCATE_DEFAULT_PTY, ALLOCATE_DEFAULT_PTY_DEFAULT);
+        heartbeatInterval = options.getInteger(HEARTBEAT_INTERVAL, HEARTBEAT_INTERVAL_DEFAULT);
         if (allocateDefaultPty) {
             logger.warn("The " + ALLOCATE_DEFAULT_PTY + " connection option has been deprecated in favour of the " + ALLOCATE_PTY + " option. See https://github.com/xebialabs/overthere#ssh_allocatePty");
         }
@@ -158,6 +149,7 @@ abstract class SshConnection extends BaseOverthereConnection {
             SSHClient client = sshClientFactory.create();
             client.setConnectTimeout(connectionTimeoutMillis);
             client.addHostKeyVerifier(new PromiscuousVerifier());
+            client.getTransport().setHeartbeatInterval(heartbeatInterval);
 
             try {
                 if (localAddress == null) {

--- a/src/main/java/com/xebialabs/overthere/ssh/SshConnectionBuilder.java
+++ b/src/main/java/com/xebialabs/overthere/ssh/SshConnectionBuilder.java
@@ -62,6 +62,16 @@ public class SshConnectionBuilder implements OverthereConnectionBuilder {
     public static final String ALLOCATE_PTY = "allocatePty";
 
     /**
+     * See <a href="https://github.com/xebialabs/overthere/blob/master/README.md#ssh_heartbeatInterval">the online documentation</a>
+     */
+    public static final String HEARTBEAT_INTERVAL = "heartbeatInterval";
+
+    /**
+     * See <a href="https://github.com/xebialabs/overthere/blob/master/README.md#ssh_heartbeatInterval">the online documentation</a>
+     */
+    public static final int HEARTBEAT_INTERVAL_DEFAULT = 0;
+
+    /**
      * See <a href="https://github.com/xebialabs/overthere/blob/master/README.md#ssh_connectionType">the online documentation</a>
      */
     public static final String CONNECTION_TYPE = "connectionType";

--- a/src/test/java/com/xebialabs/overthere/ssh/SshConnectionTest.java
+++ b/src/test/java/com/xebialabs/overthere/ssh/SshConnectionTest.java
@@ -23,6 +23,10 @@
 package com.xebialabs.overthere.ssh;
 
 import java.io.IOException;
+
+import net.schmizz.sshj.connection.ConnectionException;
+import net.schmizz.sshj.transport.Transport;
+import net.schmizz.sshj.transport.TransportException;
 import org.mockito.Matchers;
 import org.mockito.Mock;
 import org.mockito.MockitoAnnotations;
@@ -43,15 +47,9 @@ import static com.xebialabs.overthere.ConnectionOptions.OPERATING_SYSTEM;
 import static com.xebialabs.overthere.ConnectionOptions.PASSWORD;
 import static com.xebialabs.overthere.ConnectionOptions.USERNAME;
 import static com.xebialabs.overthere.OperatingSystemFamily.UNIX;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.ALLOCATE_DEFAULT_PTY;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.ALLOCATE_PTY;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.CONNECTION_TYPE;
-import static com.xebialabs.overthere.ssh.SshConnectionBuilder.PRIVATE_KEY_FILE;
+import static com.xebialabs.overthere.ssh.SshConnectionBuilder.*;
 import static com.xebialabs.overthere.ssh.SshConnectionType.SFTP;
-import static org.mockito.Matchers.anyMap;
-import static org.mockito.Matchers.anyString;
-import static org.mockito.Matchers.eq;
-import static org.mockito.Matchers.isA;
+import static org.mockito.Matchers.*;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.times;
@@ -68,14 +66,21 @@ public class SshConnectionTest {
 
     private ConnectionOptions connectionOptions;
 
+    private Session session;
+    private Transport transport;
+
     @BeforeMethod
-    public void init() {
+    public void init() throws ConnectionException, TransportException {
         MockitoAnnotations.initMocks(this);
         connectionOptions = new ConnectionOptions();
         connectionOptions.set(CONNECTION_TYPE, SFTP);
         connectionOptions.set(OPERATING_SYSTEM, UNIX);
         connectionOptions.set(ADDRESS, "nowhere.example.com");
         connectionOptions.set(USERNAME, "some-user");
+        session = mock(Session.class);
+        when(client.startSession()).thenReturn(session);
+        transport = mock(Transport.class);
+        when(client.getTransport()).thenReturn(transport);
     }
 
     @Test
@@ -117,8 +122,6 @@ public class SshConnectionTest {
 
     @Test
     public void shouldNotAllocateDefaultPty() throws IOException {
-        Session session = mock(Session.class);
-        when(client.startSession()).thenReturn(session);
         connectionOptions.set(ALLOCATE_DEFAULT_PTY, false);
 
         SshConnection connection = newConnectionWithClient(client);
@@ -131,8 +134,6 @@ public class SshConnectionTest {
 
     @Test
     public void shouldAllocateDefaultPty() throws IOException {
-        Session session = mock(Session.class);
-        when(client.startSession()).thenReturn(session);
         connectionOptions.set(ALLOCATE_DEFAULT_PTY, true);
 
         SshConnection connection = newConnectionWithClient(client);
@@ -146,8 +147,6 @@ public class SshConnectionTest {
     @SuppressWarnings("unchecked")
     @Test
     public void shouldAllocatePty() throws IOException {
-        Session session = mock(Session.class);
-        when(client.startSession()).thenReturn(session);
         connectionOptions.set(ALLOCATE_PTY, "xterm:132:50:264:100");
 
         SshConnection connection = newConnectionWithClient(client);
@@ -161,8 +160,6 @@ public class SshConnectionTest {
     @SuppressWarnings("unchecked")
     @Test
     public void allocatePtyShouldOverrideAllocateDefaultPty() throws IOException {
-        Session session = mock(Session.class);
-        when(client.startSession()).thenReturn(session);
         connectionOptions.set(ALLOCATE_PTY, "xterm:132:50:264:100");
         connectionOptions.set(ALLOCATE_DEFAULT_PTY, true);
 
@@ -174,6 +171,25 @@ public class SshConnectionTest {
         verify(session, times(0)).allocateDefaultPTY();
         connection.close();
 
+    }
+
+    @Test
+    public void shouldUseHeartbeat() throws IOException {
+        connectionOptions.set(HEARTBEAT_INTERVAL, 30);
+        SshConnection connection = newConnectionWithClient(client);
+        connection.connect();
+
+        verify(transport, times(1)).setHeartbeatInterval(30);
+        connection.close();
+    }
+
+    @Test
+    public void shouldUseDefaultHeartbeat() throws IOException {
+        SshConnection connection = newConnectionWithClient(client);
+        connection.connect();
+
+        verify(transport, times(1)).setHeartbeatInterval(0);
+        connection.close();
     }
 
     private SshConnection newConnectionWithClient(SSHClient client) {

